unstructured/documents/elements.py

```diff
@@ -509,8 +510,19 @@ def assign_and_map_hash_ids(elements: list[Element]) -> list[Element]:
     Returns:
         List of updated Element objects with hashes for `id` and `parent_id`.
     """
-    old_to_new_mapping = {e.id: e.id_to_hash() for e in elements}
-
+    # -- generate sequence number for each element on a page --
+    page_numbers = [e.metadata.page_number for e in elements]
+    page_seq_pairs = [
+        seq_on_page for page, group in groupby(page_numbers) for seq_on_page, _ in enumerate(group)
+    ]
```

Comment:
I think this is going to require more testing. The following cases must be verified by tests:
- All `element.metadata.page_number` values are `None`, such as for file-formats that do not support pagination. That would include sometimes-HTML, sometimes-DOCX, probably CSV, maybe TXT, and others.
- Some `element.metadata.page_number` values are `None`, like `PageBreak("")` and maybe `PageNumber()`.

The latter case could cause false groups/restarts.

I think this is the logic we want:
```python
    def iter_element_sequence_number_pairs():
        current_page_number, sequence_number = 1, 1
        for e in elements:
            page_number = e.metadata.page_number
            # -- An occasional element such as PageBreak() that has no page-number assumes the current
            # -- page number. An element stream with no page-numbers gets page_number=1 for each
            # -- of its elements.
            if page_number and page_number != current_page_number:
                current_page_number = page_number
                sequence_number = 1
            yield e, sequence_number
            sequence_number += 1
```

As you know I usually have a preference for comprehensions, but given the special-case logic required here I'm inclined to implement it as a generator like this (can be a function internal to `assign_and_map_hash_ids()`). That way it can be tweaked and commented as required to deal with exception cases.